### DISCUSSION TITLE
enum from/to string

### DISCUSSION THREAD

**Discussion by radare** [vlang/v#11655] [Ideas]
**V version:** V 0.1.25 876b73f
**OS:** macOS 10.15

**What did you do?**

enum foo {bar}
eprintln(foo.bar.str)

**What did you expect to see?**

all enum types should provide a compiler helper to resolve an enum type by string or get the string representation of the enum type. it is so tedious to do that in GLIB programs for example, as well as in C, and i think V should provide such helpers otu of the box to avoid the ugly spaguetti in vlib/compiler/main.v - os_from_string(os string) pref.OS
 
**What did you see instead?**

**Comment by medvednikov** [comment]
Yes, this will be implemented.

**Comment by RiordanIX** [comment]
https://github.com/vlang/v/pull/4180 implemented this, did it not? If so, this can be closed.

**Comment by medvednikov** [comment]
This issue is also about the opposite: string to enum.

**Comment by Hayao0819** [comment]
I arrived at this discussion from Google. Haven't enums been implemented from strings yet?

**Reply by xqliang** [reply]
yes, you can use the `from_string` and `str` method:
```v
enum Lang {
  en
}
assert Lang.from_string('en')? == Lang.en
assert Lang.en.str() == 'en'
```

BTW, the generic `from` is [on the way too](https://github.com/vlang/v/pull/20411):
```v
enum Lang {
  en
}
assert Lang.from('en')? == Lang.en
assert Lang.from(0)? == Lang.en
```

**Reply by mac-hel** [reply]
Behavior is inconsistent in my opinion
`.from_string()` returns Option while `.from()` returns Result

I think Option is better, I want `operation` to have `?Operation' type:
```v
enum Operation {
	get_area
	get_sector
}

fn main() {
	operation_name := ''

        // Option
	operation := ?Operation(Operation.from_string(operation_name))

        // Result - less concise
	mut operation2 := ?Operation(none)
	if r := Operation.from(operation_name) {
		operation2 = r
	}

        // Impossible
	operation3 := Operation.from(operation_name) or { ?Operation(none) }
}
```

**Reply by JalonSolov** [reply]
`.from` gives a Result because `Operation.from('fizzle')`, for example, is an error, not `none`.

If it was a query, then yes, an Option would be appropriate.  It is not, it is "create an enum **from** this string", but there is no such enum with that name.  Since it cannot succeed that way, it is an error.
